File: engine/src/providers/ollama.rs
use crate::config::ProviderConfig;
use crate::types::*;
use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Debug, Serialize, Deserialize)]
pub struct OllamaModel {
    pub name: String,
    pub size: u64,
    #[serde(default)]
    pub details: OllamaModelDetails,
}

#[derive(Debug, Default, Serialize, Deserialize)]
pub struct OllamaModelDetails {
    #[serde(default)]
    pub family: Option<String>,
}

pub async fn call_chat(
    client: &reqwest::Client,
    provider: &ProviderConfig,
    req: &ChatRequest,
) -> Result<ChatResponse, String> {
    let url = format!("{}/api/chat", provider.base_url);

    let body = serde_json::json!({
        "model": req.model,
        "stream": false,
        "messages": req.messages,
        "options": {
            "temperature": req.temperature.unwrap_or(0.0),
        },
    });

    let resp = client
        .post(&url)
        .header("Content-Type", "application/json")
        .json(&body)
        .send()
        .await
        .map_err(|e| format!("request failed: {e}"))?;

    if !resp.status().is_success() {
        let status = resp.status().as_u16();
        let text = resp.text().await.unwrap_or_default();
        return Err(format!("HTTP {status}: {text}"));
    }

    let data: Value = resp.json().await.map_err(|e| format!("json parse: {e}"))?;

    let content = data["message"]["content"]
        .as_str()
        .unwrap_or("")
        .to_string();
    let thinking = data["message"]["thinking"]
        .as_str()
        .map(|s| s.to_string());

    Ok(ChatResponse {
        id: format!("chatcmpl-{}", chrono::Utc::now().timestamp_millis()),
        object: "chat.completion".into(),
        created: chrono::Utc::now().timestamp() as u64,
        model: data["model"].as_str().unwrap_or(&req.model).into(),
        choices: vec![ChatChoice {
            index: 0,
            message: ChatMessage {
                role: "assistant".into(),
                content,
                name: None,
                reasoning: thinking,
            },
            finish_reason: "stop".into(),
        }],
        usage: Some(Usage {
            prompt_tokens: data["prompt_eval_count"].as_u64().unwrap_or(0) as u32,
            completion_tokens: data["eval_count"].as_u64().unwrap_or(0) as u32,
            total_tokens: (data["prompt_eval_count"].as_u64().unwrap_or(0)
                + data["eval_count"].as_u64().unwrap_or(0)) as u32,
        }),
        opseeq: OpseeqMeta {
            provider: String::new(),
            latency_ms: 0,
            trace_id: String::new(),
        },
    })
}

pub async fn list_models(
    client: &reqwest::Client,
    ollama_url: &str,
) -> Result<Vec<OllamaModel>, String> {
    if ollama_url.is_empty() {
        return Ok(vec![]);
    }
    let url = format!("{}/api/tags", ollama_url);
    let resp = client
        .get(&url)
        .timeout(std::time::Duration::from_secs(3))
        .send()
        .await
        .map_err(|e| format!("ollama tags: {e}"))?;

    let data: Value = resp.json().await.map_err(|e| format!("json: {e}"))?;
    let models: Vec<OllamaModel> = data["models"]
        .as_array()
        .unwrap_or(&vec![])
        .iter()
        .filter_map(|m| serde_json::from_value(m.clone()).ok())
        .collect();

    Ok(models)
}

File: copilot/models/ollama.ts
// models/ollama.ts — Local Ollama adapter (chat + embeddings).
import { env } from "./env.ts";
import type {
  ChatCompletion,
  ChatRequest,
  ChatResponse,
  EmbeddingRequest,
  EmbeddingResponse,
  ProviderId,
} from "./types.ts";

export class OllamaProvider implements ChatCompletion {
  readonly id: ProviderId = "ollama";
  readonly baseUrl: string;

  constructor() {
    this.baseUrl = env("OLLAMA_BASE_URL", "http://127.0.0.1:11434");
  }

  isReady(): boolean {
    return true; // Local server presence is checked at request time.
  }

  async chat(req: ChatRequest): Promise<ChatResponse> {
    const url = `${this.baseUrl}/api/chat`;
    const t0 = performance.now();
    const res = await fetch(url, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({
        model: req.model,
        messages: req.messages,
        stream: false,
        options: {
          temperature: req.temperature ?? 0.2,
          num_predict: req.max_tokens ?? 2048,
          ...(req.extra ?? {}),
        },
      }),
    });
    const latency_ms = Math.round(performance.now() - t0);
    if (!res.ok) {
      throw new Error(`ollama: HTTP ${res.status} ${await res.text()}`);
    }
    const data = (await res.json()) as {
      message?: { content: string };
      eval_count?: number;
      prompt_eval_count?: number;
    };
    const completion_tokens = data.eval_count ?? 0;
    const prompt_tokens = data.prompt_eval_count ?? 0;
    return {
      id: `ollama-${Date.now()}`,
      model: req.model,
      provider: "ollama",
      content: data.message?.content ?? "",
      usage: {
        prompt_tokens,
        completion_tokens,
        total_tokens: prompt_tokens + completion_tokens,
      },
      latency_ms,
      raw: data,
    };
  }

  async embed(req: EmbeddingRequest): Promise<EmbeddingResponse> {
    const url = `${this.baseUrl}/api/embed`;
    const inputs = Array.isArray(req.input) ? req.input : [req.input];
    const res = await fetch(url, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({ model: req.model, input: inputs }),
    });
    if (!res.ok) {
      throw new Error(`ollama embed: HTTP ${res.status} ${await res.text()}`);
    }
    const data = (await res.json()) as { embeddings?: number[][] };
    return {
      model: req.model,
      provider: "ollama",
      vectors: data.embeddings ?? [],
    };
  }
}
